From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Fri, 18 Feb 2022 09:49:36 +0100
Subject: Bridge: Add per port mac learning limit

This allows to configure a limit of the maximal mac addresses which can
be learned per port.
---
 bridge/link.c                | 14 ++++++++++++++
 include/uapi/linux/if_link.h |  1 +
 2 files changed, 15 insertions(+)

diff --git a/bridge/link.c b/bridge/link.c
index 9dd7475d..a967c3b1 100644
--- a/bridge/link.c
+++ b/bridge/link.c
@@ -123,6 +123,11 @@ static void print_protinfo(FILE *fp, struct rtattr *attr)
 				   "cost %u ",
 				   rta_getattr_u32(prtb[IFLA_BRPORT_COST]));
 
+		if (prtb[IFLA_BRPORT_LEARNING_LIMIT])
+			print_int(PRINT_ANY, "learning_limit",
+				   "learning_limit %d ",
+				   rta_getattr_s32(prtb[IFLA_BRPORT_LEARNING_LIMIT]));
+
 		if (!show_details)
 			return;
 
@@ -283,6 +288,7 @@ static void usage(void)
 {
 	fprintf(stderr,
 		"Usage: bridge link set dev DEV [ cost COST ] [ priority PRIO ] [ state STATE ]\n"
+		"                               [ learning_limit ] [ LEARNING_LIMIT ]\n"
 		"                               [ guard {on | off} ]\n"
 		"                               [ hairpin {on | off} ]\n"
 		"                               [ fastleave {on | off} ]\n"
@@ -340,6 +346,7 @@ static int brlink_modify(int argc, char **argv)
 	__s8 root_block = -1;
 	__u32 cost = 0;
 	__s16 priority = -1;
+	__s32 learning_limit = -2;
 	__s8 state = -1;
 	__s16 mode = -1;
 	__u16 flags = 0;
@@ -413,6 +420,9 @@ static int brlink_modify(int argc, char **argv)
 		} else if (strcmp(*argv, "priority") == 0) {
 			NEXT_ARG();
 			priority = atoi(*argv);
+		} else if (strcmp(*argv, "learning_limit") == 0) {
+			NEXT_ARG();
+			learning_limit = atoi(*argv);
 		} else if (strcmp(*argv, "state") == 0) {
 			NEXT_ARG();
 			char *endptr;
@@ -538,6 +548,10 @@ static int brlink_modify(int argc, char **argv)
 	if (priority >= 0)
 		addattr16(&req.n, sizeof(req), IFLA_BRPORT_PRIORITY, priority);
 
+	if (learning_limit >= -1)
+		addattr32(&req.n, sizeof(req), IFLA_BRPORT_LEARNING_LIMIT,
+			  learning_limit);
+
 	if (state >= 0)
 		addattr8(&req.n, sizeof(req), IFLA_BRPORT_STATE, state);
 
diff --git a/include/uapi/linux/if_link.h b/include/uapi/linux/if_link.h
index d61bd32d..6aeb9732 100644
--- a/include/uapi/linux/if_link.h
+++ b/include/uapi/linux/if_link.h
@@ -567,6 +567,7 @@ enum {
 	IFLA_BRPORT_MAB,
 	IFLA_BRPORT_MCAST_N_GROUPS,
 	IFLA_BRPORT_MCAST_MAX_GROUPS,
+	IFLA_BRPORT_LEARNING_LIMIT,
 	__IFLA_BRPORT_MAX
 };
 #define IFLA_BRPORT_MAX (__IFLA_BRPORT_MAX - 1)
